src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
+                        if (!missed) {
+                            emitting = false;
+                            skipFinal = true;
+                            return;
+                        }
+                        a = fastArray;
+                        s = fastSize;
+                        fastArray = null;
+                        fastSize = 0;
+                        missed = false;
                     }
-                    if ((wip == 0 && completed) || (wip < 0)) {
-                        sendOnComplete = true;
+                    for (int i = 0; i < s; i++) {
+                        Object v = a[i];
+                        if (v == REPORT_ERROR) {
+                            reportError();
+                            skipFinal = true;
+                            return;
+                        } else
+                        if (v == REPORT_COMPLETED) {
+                            child.onCompleted();
+                            skipFinal = true;
+                            return;
+                        }
+                        @SuppressWarnings("unchecked")
+                        T v2 = (T)v;
+                        child.onNext(v2);
                     }
                 }
-                if (sendOnComplete) {
-                    drainAndComplete();
+            } finally {
+                if (!skipFinal) {
+                    synchronized (this) {
+                        emitting = false;
+                    }
                 }
-            } else {
-                actual.onError(e);
             }
         }
-
-        @Override
-        public void onCompleted() {
-            boolean c = false;
+        void fastEmitError() {
             synchronized (this) {
-                completed = true;
-                if (wip == 0) {
-                    c = true;
+                if (emitting) {
+                    Object[] o = new Object[1];
+                    o[0] = REPORT_ERROR;
+                    fastArray = o;
+                    fastSize = 1;
+                    missed = true;
+                    return;
                 }
+                emitting = true;
             }
-            if (c) {
-                // complete outside of lock
-                drainAndComplete();
-            }
+            reportError();
+            // terminal event, no need to loop or stop emitting
         }
-
-        void completeInner(InnerSubscriber<T> s) {
-            boolean sendOnComplete = false;
+        void fastEmitCompleted() {
             synchronized (this) {
-                wip--;
-                if (wip == 0 && completed) {
-                    sendOnComplete = true;
+                if (emitting) {
+                    int s = fastSize;
+                    Object[] o = fastArray;
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
```

Comment:
This array management stuff is repeated at least twice, and should probably be used the same in all three `fastNext/Error/Completed` methods. 

Seems like a good utility to extract out as it is verbose and should always be changed the same in all of them if fixes need to occur.